src/database/test/DatabaseTests.cpp

```diff
@@ -349,3 +355,430 @@ TEST_CASE("schema test", "[db]")
+            make_optional<Liabilities>(Liabilities{5, 8}),
+            nullopt<Liabilities>()};
+
+        // Pair up each of the optional liabilities in tlOptLiabilities with a
+        // new valid account.
+        AccOptLiabilitiesVec accOptLiabilitiesVec;
+        std::transform(accOptLiabilities.begin(), accOptLiabilities.end(),
+                       std::back_inserter(accOptLiabilitiesVec),
+                       [](OptLiabilities const& aol) {
+                           return AccOptLiabilitiesPair(
+                               LedgerTestUtils::generateValidAccountEntry(),
+                               aol);
+                       });
+
+        // Pair up each of the optional liabilities in tlOptLiabilities with a
+        // new valid trustline.
+        TLOptLiabilitiesVec tlOptLiabilitiesVec;
+        std::transform(tlOptLiabilities.begin(), tlOptLiabilities.end(),
+                       std::back_inserter(tlOptLiabilitiesVec),
+                       [](OptLiabilities const& tlol) {
+                           return TLOptLiabilitiesPair(
+                               LedgerTestUtils::generateValidTrustLineEntry(),
+                               tlol);
+                       });
+
+        // Create a data entry to test that its extension and ledger
+        // entry extension have the default (empty-XDR-union) values.
+        auto de = LedgerTestUtils::generateValidDataEntry();
+
+        // Create an offer entry to test that its extension and ledger
+        // entry extension have the default (empty-XDR-union) values.
+        auto oe = LedgerTestUtils::generateValidOfferEntry();
+
+        // Create the application, with the code above that inserts old-schema
+        // accounts and trustlines into the database injected between database
+        // creation and upgrade.
+        Config const& cfg = getTestConfig(0, dbMode);
+        VirtualClock clock;
+        Application::pointer app =
+            createTestApplication<SchemaUpgradeTestApplication,
+                                  SchemaUpgradeTestApplication::PreUpgradeFunc>(
+                clock, cfg,
+                [prepOldSchemaDB, accOptLiabilitiesVec, tlOptLiabilitiesVec, de,
+                 oe](SchemaUpgradeTestApplication& sapp) {
+                    prepOldSchemaDB(sapp, accOptLiabilitiesVec,
+                                    tlOptLiabilitiesVec, de, oe);
+                });
+        app->start();
+
+        // Validate that the accounts and trustlines have the expected
+        // liabilities and ledger entry extensions now that the database upgrade
+        // has completed.
+
+        LedgerTxn ltx(app->getLedgerTxnRoot());
+
+        for (auto aol : accOptLiabilitiesVec)
+        {
+            LedgerKey key;
+            key.type(ACCOUNT);
+            key.account().accountID = aol.first.accountID;
+            auto acc = ltx.load(key);
+            REQUIRE(acc.current().data.type() == ACCOUNT);
```

Comment:
Why effectively write your own comparison operator? Use the built-in XDR comparison operators for `LedgerEntry` instead. It'll save a lot of effort, reduce the risk of missing something, and improve clarity. (This applies in multiple places)

Reply:
I wanted to test with XDR's ==, but as the code was, that comparison didn't hold!  The extensions could have differed, because the loaded one could have been changed by the optional Liabilities paired with the given entries.  However, now that I've implemented your suggestion of just using the extensions themselves to represent the optional<>, the operator== works.  So your earlier idea led to a cascade of cleanups.  Thanks!